File: htorrent/sha1hex.py
#!/usr/bin/env python3
import urllib.parse
import hashlib
import sys

def main():
    file_name = sys.argv[1]
    with open(file_name, 'rb') as f:
        file_contents = f.read()
    sha1 = hashlib.sha1(file_contents)
    print(urllib.parse.urlencode({'a': sha1.digest()}))

if __name__ == '__main__':
    main()

File: htorrent/sha1hex.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.2 script

import qualified System.Environment as Sys
import qualified Data.ByteString as BS
import qualified Data.ByteString.UTF8 as UTF8
import qualified Crypto.Hash as C


main :: IO ()
main = do
  args <- Sys.getArgs
  file <- BS.readFile $ head args
  print $ (C.hash file :: C.Digest C.SHA1)

